File: push-be/internal/domain/usecase/evidence.go
package usecase

import (
	"context"
	"encoding/json"
	"errors"
	"time"

	"github.com/google/uuid"

	"push-be/internal/domain"
)

type EvidenceService struct {
	evidence domain.EvidenceStore
	ops      domain.OperationStore
	uow      domain.UnitOfWork
}

func NewEvidenceService(evidence domain.EvidenceStore, ops domain.OperationStore, uow domain.UnitOfWork) *EvidenceService {
	return &EvidenceService{evidence: evidence, ops: ops, uow: uow}
}

type CreateEvidenceInput struct {
	Kind         domain.EvidenceKind
	Title        string
	SourceText   string
	SourceURL    *string
	Skills       []string
	SupersedesID *uuid.UUID
}

func validateEvidenceInput(in *CreateEvidenceInput) *domain.Error {
	if !domain.ValidEvidenceKind(in.Kind) {
		return domain.ValidationField("kind", "unsupported kind")
	}
	if len(in.Title) < 1 || len(in.Title) > 200 {
		return domain.ValidationField("title", "title must be 1-200 characters")
	}
	if n := len(in.SourceText); n < 1 || n > 100000 {
		return domain.ValidationField("sourceText", "sourceText must be 1-100000 characters")
	}
	if len(in.Skills) > 100 {
		return domain.ValidationField("skills", "at most 100 skills")
	}
	for _, sk := range in.Skills {
		if len(sk) < 1 || len(sk) > 100 {
			return domain.ValidationField("skills", "each skill must be 1-100 characters")
		}
	}
	if in.SourceURL != nil {
		if err := domain.ValidateHTTPSURL(*in.SourceURL); err != nil {
			return domain.ValidationField("sourceUrl", "must be an http(s) URL without credentials")
		}
	}
	return nil
}

func (s *EvidenceService) List(ctx context.Context, userID uuid.UUID, f domain.EvidenceFilter) (domain.Page[domain.CareerEvidence], error) {
	limit := f.Page.EffectiveLimit()
	items, err := s.evidence.List(ctx, userID, f, limit+1)
	if err != nil {
		return domain.Page[domain.CareerEvidence]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(e domain.CareerEvidence) domain.Cursor {
		return domain.Cursor{CreatedAt: e.CreatedAt, ID: e.ID}
	}), nil
}

func (s *EvidenceService) Get(ctx context.Context, userID, id uuid.UUID) (*domain.CareerEvidence, error) {
	e, err := s.evidence.Get(ctx, userID, id)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	return e, nil
}

func (s *EvidenceService) Create(ctx context.Context, userID uuid.UUID, in CreateEvidenceInput) (*domain.CareerEvidence, error) {
	if err := validateEvidenceInput(&in); err != nil {
		return nil, err
	}
	now := time.Now().UTC()
	e := &domain.CareerEvidence{
		ID: uuid.New(), UserID: userID, Revision: 1,
		Kind: in.Kind, Title: in.Title, SourceText: in.SourceText, SourceURL: in.SourceURL,
		Skills: in.Skills, VerificationStatus: domain.VerificationUserProvided,
		Provenance:   domain.Provenance{ContentHash: domain.HashBytes([]byte(in.SourceText))},
		SupersedesID: in.SupersedesID, CreatedAt: now, UpdatedAt: now,
	}
	if e.Skills == nil {
		e.Skills = []string{}
	}
	if in.SupersedesID != nil {
		if _, err := s.evidence.Get(ctx, userID, *in.SupersedesID); err != nil {
			return nil, domain.ValidationField("supersedesId", "referenced evidence not found")
		}
	}
	if err := s.evidence.Create(ctx, e); err != nil {
		return nil, domain.Internal()
	}
	return e, nil
}

func (s *EvidenceService) Archive(ctx context.Context, userID, id uuid.UUID, expectedRevision int64) (*domain.CareerEvidence, error) {
	var out *domain.CareerEvidence
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		e, err := s.evidence.Get(ctx, userID, id)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if e.Archived {
			out = e
			return nil
		}
		e.Archived = true
		e.UpdatedAt = time.Now().UTC()
		if err := s.evidence.Update(ctx, e, expectedRevision); err != nil {
			return mapRevisionErr(err)
		}
		e.Revision = expectedRevision + 1
		out = e
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

func mapRevisionErr(err error) error {
	if errors.Is(err, domain.ErrNotFound) {
		return domain.NotFound()
	}
	if c, ok := domain.AsConflict(err); ok {
		return domain.RevisionConflict(c.Current)
	}
	return err
}

type ImportInput struct {
	SourceID    *uuid.UUID
	Text        string
	SourceURL   *string
	ContentHash string
	Format      string
}

func ValidImportFormat(f string) bool {
	switch f {
	case "TEXT", "MARKDOWN", "PDF", "DOCX", "GITHUB":
		return true
	}
	return false
}

func (s *EvidenceService) Import(ctx context.Context, userID uuid.UUID, in ImportInput) (*domain.Operation, error) {
	if !ValidImportFormat(in.Format) {
		return nil, domain.ValidationField("format", "unsupported format")
	}
	if in.Format == "GITHUB" {
		return nil, domain.IntegrationRequired("GitHub collection requires a connected integration")
	}
	if in.SourceURL != nil {
		if err := domain.ValidateHTTPSURL(*in.SourceURL); err != nil {
			return nil, domain.ValidationField("sourceUrl", "must be an http(s) URL without credentials")
		}
	}
	if in.SourceID != nil {
		return nil, domain.ValidationField("sourceId", "stored sources are not supported by this build; pass extracted text")
	}
	now := time.Now().UTC()
	payload, _ := json.Marshal(map[string]any{
		"text": in.Text, "sourceUrl": in.SourceURL, "contentHash": in.ContentHash, "format": in.Format,
	})
	op := &domain.Operation{
		ID: uuid.New(), UserID: userID, Type: domain.OpEvidenceImport,
		CreatedAt: now, UpdatedAt: now, PendingPayload: payload,
	}
	s.progressImport(op)
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		if err := s.persistImportResult(ctx, op); err != nil {
			return err
		}
		return s.ops.Create(ctx, op)
	})
	if err != nil {
		return nil, domain.Internal()
	}
	return op, nil
}

func (s *EvidenceService) persistImportResult(ctx context.Context, op *domain.Operation) error {
	if op.Status != domain.OpSucceeded || op.Result == nil {
		return nil
	}
	var res struct {
		Evidence []domain.CareerEvidence `json:"evidence"`
	}
	if err := json.Unmarshal(op.Result.Value, &res); err != nil || len(res.Evidence) == 0 {
		return nil
	}
	created := &res.Evidence[0]
	if err := s.evidence.Create(ctx, created); err != nil {
		return err
	}
	v, _ := json.Marshal(map[string]any{"evidence": []any{created}, "warnings": []any{}})
	op.Result = &domain.OperationResult{Kind: string(domain.OpEvidenceImport), Value: v}
	return nil
}

func (s *EvidenceService) progressImport(op *domain.Operation) {
	var p struct {
		Text        string  `json:"text"`
		SourceURL   *string `json:"sourceUrl"`
		ContentHash string  `json:"contentHash"`
		Format      string  `json:"format"`
		Kind        string  `json:"kind"`
		Title       string  `json:"title"`
	}
	_ = json.Unmarshal(op.PendingPayload, &p)
	missing := []domain.InputField{}
	if p.Text == "" {
		missing = append(missing, domain.InputField{Name: "text", Label: "Extracted source text", Type: "TEXT"})
	}
	if !domain.ValidEvidenceKind(domain.EvidenceKind(p.Kind)) {
		missing = append(missing, domain.InputField{Name: "kind", Label: "Evidence kind (RESUME|CAREER|EDUCATION|SKILL|PROJECT)", Type: "TEXT"})
	}
	if p.Title == "" {
		missing = append(missing, domain.InputField{Name: "title", Label: "Evidence title", Type: "TEXT"})
	}
	now := time.Now().UTC()
	if len(missing) > 0 {
		op.Status = domain.OpNeedsInput
		op.InputRequest = &domain.InputRequest{
			Code:    "IMPORT_DETAILS_REQUIRED",
			Message: "text, kind and title are required to create evidence",
			Fields:  missing,
		}
		op.UpdatedAt = now
		return
	}
	evidence := &domain.CareerEvidence{
		ID: uuid.New(), UserID: op.UserID, Revision: 1,
		Kind: domain.EvidenceKind(p.Kind), Title: p.Title, SourceText: p.Text,
		SourceURL: p.SourceURL, Skills: []string{},
		VerificationStatus: domain.VerificationUserProvided,
		Provenance: domain.Provenance{
			ContentHash:    firstNonEmpty(p.ContentHash, domain.HashBytes([]byte(p.Text))),
			SourceLocation: &domain.SourceLocation{Start: 0, End: domain.CodePointLen(p.Text), Unit: "CODE_POINT"},
		},
		CreatedAt: now, UpdatedAt: now,
	}
	op.Result = evidenceResult(evidence)
	op.Status = domain.OpSucceeded
	op.UpdatedAt = now
}

func evidenceResult(e *domain.CareerEvidence) *domain.OperationResult {
	v, _ := json.Marshal(map[string]any{
		"evidence": []any{e}, "warnings": []any{},
	})
	return &domain.OperationResult{Kind: string(domain.OpEvidenceImport), Value: v}
}

func firstNonEmpty(a, b string) string {
	if a != "" {
		return a
	}
	return b
}

func (s *EvidenceService) CompleteImportInput(ctx context.Context, op *domain.Operation, fields map[string]string) error {
	var p map[string]any
	if err := json.Unmarshal(op.PendingPayload, &p); err != nil {
		p = map[string]any{}
	}
	allowed := map[string]bool{}
	for _, f := range op.InputRequest.Fields {
		allowed[f.Name] = true
	}
	for k, v := range fields {
		if !allowed[k] {
			return domain.ValidationField("fields."+k, "field not requested")
		}
		p[k] = v
	}
	op.PendingPayload, _ = json.Marshal(p)
	s.progressImport(op)
	return s.persistImportResult(ctx, op)
}

File: push-be/app/domain/services/evidence.py
from __future__ import annotations
from datetime import datetime, timezone
from typing import Optional
from uuid import UUID, uuid4

from app.db import DB, NotFoundError
from app.domain import entities as ent
from app.domain.errors import (
    internal, map_revision_err, not_found, validation_field, integration_required,
)
from app.domain.validators import code_point_len, hash_bytes, validate_https_url
from app.infrastructure.store_evidence import EvidenceStore
from app.infrastructure.store_operations import OperationStore
from app.jsonutil import to_jsonable


def _now() -> datetime:
    return datetime.now(timezone.utc)


IMPORT_FORMATS = {"TEXT", "MARKDOWN", "PDF", "DOCX", "GITHUB"}


def validate_evidence_input(kind: str, title: str, source_text: str,
                            source_url: Optional[str], skills: list[str],
                            supersedes_id: Optional[UUID]) -> None:
    if not ent.valid_evidence_kind(kind):
        raise validation_field("kind", "unsupported kind")
    if not 1 <= len(title) <= 200:
        raise validation_field("title", "title must be 1-200 characters")
    if not 1 <= len(source_text) <= 100000:
        raise validation_field("sourceText", "sourceText must be 1-100000 characters")
    if len(skills) > 100:
        raise validation_field("skills", "at most 100 skills")
    if any(not 1 <= len(s) <= 100 for s in skills):
        raise validation_field("skills", "each skill must be 1-100 characters")
    if source_url is not None:
        try:
            validate_https_url(source_url)
        except ValueError:
            raise validation_field("sourceUrl", "must be an http(s) URL without credentials")


class EvidenceService:
    def __init__(self, db: DB):
        self.db = db
        self.evidence = EvidenceStore(db)
        self.ops = OperationStore(db)

    async def list(self, user_id: UUID, kind: Optional[str], query: str, page) -> ent.Page:
        try:
            return await self.evidence.list(user_id, kind, query, page)
        except Exception:
            raise internal()

    async def get(self, user_id: UUID, id_: UUID) -> ent.CareerEvidence:
        try:
            return await self.evidence.get(user_id, id_)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()

    async def create(self, user_id: UUID, kind: str, title: str, source_text: str,
                     source_url: Optional[str], skills: list[str],
                     supersedes_id: Optional[UUID]) -> ent.CareerEvidence:
        validate_evidence_input(kind, title, source_text, source_url, skills or [], supersedes_id)
        now = _now()
        e = ent.CareerEvidence(
            id=uuid4(), user_id=user_id, revision=1, kind=kind, title=title,
            source_text=source_text, source_url=source_url, skills=skills or [],
            verification_status=ent.VERIFICATION_USER_PROVIDED,
            provenance=ent.Provenance(content_hash=hash_bytes(source_text.encode())),
            supersedes_id=supersedes_id, created_at=now, updated_at=now)
        if supersedes_id is not None:
            try:
                await self.evidence.get(user_id, supersedes_id)
            except NotFoundError:
                raise validation_field("supersedesId", "referenced evidence not found")
        try:
            await self.evidence.create(e)
        except Exception:
            raise internal()
        return e

    async def archive(self, user_id: UUID, id_: UUID, expected: int) -> ent.CareerEvidence:
        out = None

        async def work():
            nonlocal out
            try:
                e = await self.evidence.get(user_id, id_)
            except NotFoundError:
                raise not_found()
            if e.archived:
                out = e
                return
            e.archived = True
            e.updated_at = _now()
            try:
                await self.evidence.update(e, expected)
            except Exception as err:
                raise map_revision_err(err)
            e.revision = expected + 1
            out = e

        await self.db.do(work)
        return out

    async def import_(self, user_id: UUID, source_id: Optional[UUID], text: str,
                      source_url: Optional[str], content_hash: str,
                      format_: str) -> ent.Operation:
        if format_ not in IMPORT_FORMATS:
            raise validation_field("format", "unsupported format")
        if format_ == "GITHUB":
            raise integration_required("GitHub collection requires a connected integration")
        if source_url is not None:
            try:
                validate_https_url(source_url)
            except ValueError:
                raise validation_field("sourceUrl", "must be an http(s) URL without credentials")
        if source_id is not None:
            raise validation_field(
                "sourceId", "stored sources are not supported by this build; pass extracted text")
        now = _now()
        op = ent.Operation(
            id=uuid4(), user_id=user_id, type=ent.OP_EVIDENCE_IMPORT, status="",
            pending_payload={"text": text, "sourceUrl": source_url,
                             "contentHash": content_hash, "format": format_},
            created_at=now, updated_at=now)
        await self._progress_import(op)

        async def work():
            await self._persist_import_result(op)
            await self.ops.create(op)

        try:
            await self.db.do(work)
        except Exception:
            raise internal()
        return op

    async def _persist_import_result(self, op: ent.Operation) -> None:
        if op.status != ent.OP_SUCCEEDED or op.result is None:
            return
        ev_list = (op.result.value or {}).get("evidence") or []
        if not ev_list:
            return
        created = ent.CareerEvidence(**ev_list[0])
        await self.evidence.create(created)
        op.result = ent.OperationResult(
            kind=ent.OP_EVIDENCE_IMPORT,
            value={"evidence": [to_jsonable(created)], "warnings": []})

    async def _progress_import(self, op: ent.Operation) -> None:
        p = op.pending_payload or {}
        missing = []
        if not p.get("text"):
            missing.append(ent.InputField(name="text", label="Extracted source text", type="TEXT"))
        if not ent.valid_evidence_kind(p.get("kind") or ""):
            missing.append(ent.InputField(
                name="kind", label="Evidence kind (RESUME|CAREER|EDUCATION|SKILL|PROJECT)",
                type="TEXT"))
        if not p.get("title"):
            missing.append(ent.InputField(name="title", label="Evidence title", type="TEXT"))
        now = _now()
        if missing:
            op.status = ent.OP_NEEDS_INPUT
            op.input_request = ent.InputRequest(
                code="IMPORT_DETAILS_REQUIRED",
                message="text, kind and title are required to create evidence",
                fields=missing)
            op.updated_at = now
            return
        evidence = ent.CareerEvidence(
            id=uuid4(), user_id=op.user_id, revision=1, kind=p["kind"], title=p["title"],
            source_text=p["text"], source_url=p.get("sourceUrl"), skills=[],
            verification_status=ent.VERIFICATION_USER_PROVIDED,
            provenance=ent.Provenance(
                content_hash=p.get("contentHash") or hash_bytes(p["text"].encode()),
                source_location=ent.SourceLocation(
                    start=0, end=code_point_len(p["text"]), unit="CODE_POINT")),
            created_at=now, updated_at=now)
        op.result = ent.OperationResult(
            kind=ent.OP_EVIDENCE_IMPORT,
            value={"evidence": [to_jsonable(evidence)], "warnings": []})
        op.status = ent.OP_SUCCEEDED
        op.updated_at = now

    async def complete_import_input(self, op: ent.Operation, fields: dict) -> None:
        p = op.pending_payload if isinstance(op.pending_payload, dict) else {}
        allowed = {f.name for f in (op.input_request.fields if op.input_request else [])}
        for k, v in fields.items():
            if k not in allowed:
                raise validation_field(f"fields.{k}", "field not requested")
            p[k] = v
        op.pending_payload = p
        await self._progress_import(op)
        await self._persist_import_result(op)
